Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs

```diff
@@ -0,0 +1,252 @@
+        {
+            try
+            {
+                _logger?.LogInformation("Ensuring activities");
+
+                var dict = InitDictionary(spActivities);
+
+                foreach (var activity in activities)
+                {
+                    UpdateDictionary(dict, activity, members);
+                }
+
+                var itemsToUpdate = dict.Where(x => x.Value.IsModified)
+                    .Select(x => new SpActivity()
+                    {
+                        Id = x.Value.ActivityId,
+                        Activities = new List<string>(x.Value.Activities),
+                        Year = x.Key.Year,
+                        Month = x.Key.Month,
+                        MentorId = x.Value.MentorId,
+                        MentorLookupId = x.Value.MentorLookupId,
+                        RootMentorId = x.Value.RootMentorId,
+                        RootMentorLookupId = x.Value.RootMentorLookupId,
+                        Paths = new List<string>(x.Value.Paths)
+                    });
+
+                return itemsToUpdate;
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Ensuring activities failed. {e.Message}");
+            }
+        }
+
+        private int? GetUserIdByEmail(string email, IEnumerable<Member> members)
+        {
+            var user = members.FirstOrDefault(x =>
+                x.MentorEmail == email || x.RootMentorEmail == email);
+
+            if (user == null) return null;
+
+            var id = user.MentorId ?? user.RootMentorId;
+            return id;
+        }
+
+        private void UpdateDictionary(Dictionary<ActivityKey, ActivityValue> dict, ActivityInfo activity, IEnumerable<Member> members)
+        {
+            var key = new ActivityKey()
+            {
+                UserId = (int)activity.UserId,
+                Year = activity.Date.Year,
+                Month = activity.Date.Month
+            };
+
+            if (dict.ContainsKey(key))
+            {
+                var value = dict[key];
+                if (!value.Activities.Contains(activity.Activity))
+                {
+                    value.Activities.Add(activity.Activity);
+                    dict[key].IsModified = true;
+                }
+
+                var set = new HashSet<string>(activity.Paths);
+                set.ExceptWith(value.Paths);
+
+                if (set.Count == 0) return;
+
+                foreach (var item in set)
+                {
+                    value.Paths.Add(item);
+                }
+                dict[key].IsModified = true;
+            }
+            else
+            {
+                var newMember = members.FirstOrDefault(x =>
+                    x.MentorId == activity.UserId || x.RootMentorId == activity.UserId);
+
+                if (newMember == null) return;
+
+                var newKey = key;
+
+                switch (activity.Activity)
```

Comment:
I don't think that this service should know this business logic. This service is not responsible for determination of activity type. Just remove this validation.